File: tools/keystore/C005Oracle.java
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.File;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.attribute.PosixFilePermissions;
import java.util.ArrayList;
import java.util.Base64;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import org.tron.common.crypto.SignInterface;
import org.tron.common.crypto.SignUtils;
import org.tron.keystore.Credentials;
import org.tron.keystore.Wallet;
import org.tron.keystore.WalletFile;
import org.tron.keystore.WalletUtils;
import org.tron.common.parameter.CommonParameter;

/** Executable C005 adapter over the pinned java-tron keystore implementation. */
public final class C005Oracle {
  private static final ObjectMapper JSON = new ObjectMapper();
  private static final String REVISION = "4a21592f95e37908b21bc3f611c6e7a1a67f09f3";
  private static final String PASSWORD = "correct horse battery staple";
  private static final String PRIVATE = "0000000000000000000000000000000000000000000000000000000000000001";
  private static final String SALT = "000102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f";
  private static final String IV = "202122232425262728292a2b2c2d2e2f";
  private static final String UUID = "00112233-4455-4677-8899-aabbccddeeff";
  private static final String EC_ADDRESS = "TMVQGm1qAQYVdetCeGRRkTWYYrLXuHK2HC";
  private static final String SM2_ADDRESS = "TUAyHbMzdj2dMAJxfKK7KrubmQdARvF1GH";

  public static void main(String[] args) throws Exception {
    if (args.length == 0 || "manifest".equals(args[0])) {
      JSON.writeValue(System.out, manifest());
    } else if ("java-create".equals(args[0])) {
      JSON.writeValue(System.out, javaCreate(engine(args)));
    } else if ("java-open".equals(args[0])) {
      String walletJson = new String(Base64.getDecoder().decode(args[2]), StandardCharsets.UTF_8);
      JSON.writeValue(System.out, javaOpen(walletJson, engine(args)));
    } else {
      throw new IllegalArgumentException("unknown mode: " + args[0]);
    }
  }

  private static boolean engine(String[] args) {
    if (args.length < 2 || !("ec".equals(args[1]) || "sm2".equals(args[1]))) {
      throw new IllegalArgumentException("engine must be ec or sm2");
    }
    boolean ec = "ec".equals(args[1]);
    configureEngine(ec);
    return ec;
  }

  private static void configureEngine(boolean ec) {
    CommonParameter.getInstance().cryptoEngine = ec ? "ECKey" : "SM2";
  }

  private static Map<String, Object> manifest() throws Exception {
    Map<String, Object> root = map();
    root.put("schema_version", 1);
    root.put("java_revision", REVISION);
    Map<String, Object> inputs = map();
    inputs.put("password", PASSWORD); inputs.put("private_key_hex", PRIVATE);
    inputs.put("salt_hex", SALT); inputs.put("iv_hex", IV); inputs.put("uuid", UUID);
    root.put("deterministic_inputs", inputs);
    List<Map<String, Object>> rows = new ArrayList<>();
    rows.add(cross("C005.CROSS.SCRYPT.EC", "scrypt", true));
    rows.add(cross("C005.CROSS.SCRYPT.SM2", "scrypt", false));
    rows.add(cross("C005.CROSS.PBKDF2.EC", "pbkdf2", true));
    rows.add(cross("C005.CROSS.PBKDF2.SM2", "pbkdf2", false));
    rows.add(pbkdf2MissingDklen());
    rows.add(error("C005.ERROR.WRONG_PASSWORD", "decrypt", "Invalid password provided"));
    rows.add(error("C005.ERROR.VERSION", "schema", "Wallet version is not supported"));
    rows.add(versionIntegerOverflow());
    rows.add(error("C005.ERROR.MISSING_CRYPTO", "schema", "Missing crypto section"));
    rows.add(error("C005.ERROR.CIPHER", "schema", "Wallet cipher is not supported"));
    rows.add(error("C005.ERROR.KDF", "schema", "KDF type is not supported"));
    rows.add(error("C005.ERROR.PRF", "decrypt", "Unsupported prf:hmac-sha1"));
    rows.add(password("C005.PASSWORD.WHITESPACE", "  six words remain  "));
    rows.add(password("C005.PASSWORD.BOM", "\ufeffsecret\r\n"));
    rows.add(password("C005.PASSWORD.EMPTY", ""));
    rows.add(password("C005.PASSWORD.SHORT", "12345"));
    rows.add(decision("C005.PASSWORD.MULTILINE", "password_file", "java_source", "Java KeystoreCliUtils.readPassword rejects password files containing multiple lines."));
    rows.add(decision("C005.PASSWORD.MALFORMED_UTF8", "password_file", "rust_security_strengthening", "Java Scanner decoding is platform-configured; Rust defines deterministic UTF-8 lossy replacement."));
    String[] javaDirect = {"DIRECT.SYMLINK", "DIRECT.UNQUOTED", "DIRECT.CORRUPT", "WRITE.CLEANUP", "WRITE.MODE"};
    for (String id : javaDirect) rows.add(filesystem("C005." + id, "java_execution", javaFilesystemProbe(id)));
    String[] strengthened = {"NEW.INJECTED", "IMPORT.DUPLICATE", "IMPORT.OVERWRITE", "LIST.CORRUPT", "LIST.BOM", "LIST.MULTILINE", "LIST.SYMLINK", "LIST.PERMISSIONS", "LIST.OWNERSHIP", "UPDATE.PASSWORD", "UPDATE.WRONG_PASSWORD", "UPDATE.DUPLICATE", "UPDATE.SYMLINK_SWAP", "WINDOWS.LIMIT"};
    for (String id : strengthened) rows.add(filesystem("C005." + id, "rust_security_strengthening", filesystemDecision(id)));
    root.put("vectors", rows);
    return root;
  }

  private static Map<String, Object> cross(String id, String kdf, boolean ec) throws Exception {
    configureEngine(ec);
    String walletJson = deterministicWallet(kdf, ec);
    Map<String, Object> opened = javaOpen(walletJson, ec);
    if (!PRIVATE.equals(opened.get("private_key_hex"))) throw new AssertionError(id + " private key");
    Map<String, Object> row = map();
    row.put("id", id); row.put("kind", "cross_open"); row.put("direction", "rust_deterministic_create_to_java_open");
    row.put("java_api", "WalletUtils.loadCredentials/Credentials.getSignInterface");
    row.put("key_engine", ec ? "secp256k1" : "sm2"); row.put("checksum_engine", ec ? "secp256k1" : "sm2");
    row.put("kdf", kdf); row.put("private_key_hex", PRIVATE); row.put("wallet_json", walletJson);
    return row;
  }

  private static Map<String, Object> pbkdf2MissingDklen() throws Exception {
    configureEngine(true);
    String id = "C005.CROSS.PBKDF2.MISSING_DKLEN.EC";
    String walletJson = deterministicWallet("pbkdf2", true).replace("\"dklen\":32,", "");
    WalletFile file = JSON.readValue(walletJson, WalletFile.class);
    WalletFile.Aes128CtrKdfParams params = (WalletFile.Aes128CtrKdfParams) file.getCrypto().getKdfparams();
    if (params.getDklen() != 0) throw new AssertionError(id + " default dklen");
    Map<String, Object> opened = javaOpen(walletJson, true);
    if (!PRIVATE.equals(opened.get("private_key_hex"))) throw new AssertionError(id + " private key");
    Map<String, Object> row = map();
    row.put("id", id); row.put("kind", "cross_open"); row.put("direction", "pbkdf2_missing_dklen_java_and_rust_open");
    row.put("java_api", "ObjectMapper.readValue/WalletFile.Aes128CtrKdfParams.getDklen/WalletUtils.loadCredentials");
    row.put("key_engine", "secp256k1"); row.put("checksum_engine", "secp256k1"); row.put("kdf", "pbkdf2");
    row.put("java_default_dklen", params.getDklen()); row.put("private_key_hex", PRIVATE); row.put("wallet_json", walletJson);
    return row;
  }

  private static Map<String, Object> javaCreate(boolean ec) throws Exception {
    SignInterface sign = SignUtils.fromPrivate(hex(PRIVATE), ec);
    WalletFile wallet = Wallet.createLight(PASSWORD, sign);
    Path dir = Files.createTempDirectory("c005-java-create-");
    try {
      Path path = dir.resolve("wallet.json");
      WalletUtils.writeWalletFile(wallet, path.toFile());
      Credentials credentials = WalletUtils.loadCredentials(PASSWORD, path.toFile(), ec);
      Map<String, Object> out = map();
      out.put("direction", "java_random_create_to_rust_open");
      out.put("java_api", "Wallet.createLight/WalletUtils.writeWalletFile/WalletUtils.loadCredentials/Credentials");
      out.put("wallet_json", JSON.writeValueAsString(wallet));
      out.put("private_key_hex", hx(credentials.getSignInterface().getPrivateKey()));
      out.put("address", credentials.getAddress());
      return out;
    } finally { deleteTree(dir); }
  }

  private static Map<String, Object> javaOpen(String walletJson, boolean ec) throws Exception {
    Path dir = Files.createTempDirectory("c005-java-open-");
    try {
      Path path = dir.resolve("wallet.json");
      Files.write(path, walletJson.getBytes(StandardCharsets.UTF_8));
      Credentials credentials = WalletUtils.loadCredentials(PASSWORD, path.toFile(), ec);
      Map<String, Object> out = map();
      out.put("private_key_hex", hx(credentials.getSignInterface().getPrivateKey()));
      out.put("address", credentials.getAddress());
      return out;
    } finally { deleteTree(dir); }
  }

  private static Map<String, Object> error(String id, String kind, String expected) throws Exception {
    configureEngine(true);
    WalletFile file = JSON.readValue(deterministicWallet(id.contains("PRF") ? "pbkdf2" : "scrypt", true), WalletFile.class);
    String password = PASSWORD;
    if (id.endsWith("WRONG_PASSWORD")) password = "incorrect";
    else if (id.endsWith("VERSION")) file.setVersion(2);
    else if (id.endsWith("MISSING_CRYPTO")) file.setCrypto(null);
    else if (id.endsWith("CIPHER")) file.getCrypto().setCipher("aes-256-gcm");
    else if (id.endsWith("KDF")) file.getCrypto().setKdf("argon2");
    else if (id.endsWith("PRF")) ((WalletFile.Aes128CtrKdfParams) file.getCrypto().getKdfparams()).setPrf("hmac-sha1");
    String actual;
    try { Credentials.create(Wallet.decrypt(password, file, true)); throw new AssertionError(id + " unexpectedly succeeded"); }
    catch (org.tron.core.exception.CipherException e) { actual = e.getMessage(); }
    if (!expected.equals(actual)) throw new AssertionError(id + ": " + actual);
    Map<String, Object> row = map(); row.put("id", id); row.put("kind", kind); row.put("expected", actual); row.put("java_api", "Wallet.decrypt"); return row;
  }
  private static Map<String, Object> versionIntegerOverflow() throws Exception {
    String walletJson = deterministicWallet("scrypt", true).replace("\"version\":3", "\"version\":4294967299");
    String actual;
    try {
      JSON.readValue(walletJson, WalletFile.class);
      throw new AssertionError("C005.ERROR.VERSION_INT_OVERFLOW unexpectedly succeeded");
    } catch (com.fasterxml.jackson.core.JsonProcessingException error) {
      actual = error.getOriginalMessage();
    }
    if (!actual.contains("4294967299") || !actual.contains("out of range of int")) {
      throw new AssertionError("C005.ERROR.VERSION_INT_OVERFLOW: " + actual);
    }
    Map<String, Object> row = map();
    row.put("id", "C005.ERROR.VERSION_INT_OVERFLOW");
    row.put("kind", "schema_json");
    row.put("expected", actual);
    row.put("java_api", "ObjectMapper.readValue/WalletFile.setVersion(int)");
    row.put("wallet_json", walletJson);
    return row;
  }


  private static void openWithPassword(String walletJson, boolean ec, String password) throws Exception {
    WalletFile file = JSON.readValue(walletJson, WalletFile.class); Credentials.create(Wallet.decrypt(password, file, ec));
  }

  private static Map<String, Object> password(String id, String input) {
    String normalized = WalletUtils.stripPasswordLine(input);
    Map<String, Object> row = map(); row.put("id", id); row.put("kind", "password"); row.put("input", input);
    row.put("normalized", normalized); row.put("valid", WalletUtils.passwordValid(normalized)); row.put("java_api", "WalletUtils.stripPasswordLine/passwordValid"); return row;
  }

  private static String javaFilesystemProbe(String id) throws Exception {
    configureEngine(true);
    Path dir = Files.createTempDirectory("c005-java-fs-");
    try {
      Path target = dir.resolve("target.json"); Files.write(target, deterministicWallet("scrypt", true).getBytes(StandardCharsets.UTF_8));
      if ("DIRECT.SYMLINK".equals(id)) { Path link=dir.resolve("link.json"); Files.createSymbolicLink(link, target.getFileName()); WalletUtils.loadCredentials(PASSWORD, link.toFile(), true); return "WalletUtils.loadCredentials follows configured symlink after warning"; }
      if ("DIRECT.UNQUOTED".equals(id)) { String loose=deterministicWallet("scrypt", true).replaceFirst("\\{\\\"address\\\"", "{address"); Files.write(target, loose.getBytes(StandardCharsets.UTF_8)); WalletUtils.loadCredentials(PASSWORD, target.toFile(), true); return "WalletUtils.loadCredentials accepts unquoted field names"; }
      if ("DIRECT.CORRUPT".equals(id)) { Files.write(target, "{".getBytes(StandardCharsets.UTF_8)); try { WalletUtils.loadCredentials(PASSWORD,target.toFile(),true); } catch (com.fasterxml.jackson.core.JsonProcessingException expected) { return "WalletUtils.loadCredentials surfaces corrupt JSON"; } throw new AssertionError(id); }
      WalletFile wallet=JSON.readValue(deterministicWallet("scrypt",true),WalletFile.class); Path output=dir.resolve("out.json");
      if ("WRITE.MODE".equals(id)) { WalletUtils.writeWalletFile(wallet,output.toFile()); String mode=PosixFilePermissions.toString(Files.getPosixFilePermissions(output)); if (!"rw-------".equals(mode)) throw new AssertionError(mode); return "WalletUtils.writeWalletFile creates owner-read/write-only output"; }
      Files.createDirectory(output); try { WalletUtils.writeWalletFile(wallet,output.toFile()); } catch (Exception expected) { try (java.util.stream.Stream<Path> paths=Files.list(dir)) { if (paths.anyMatch(p->p.getFileName().toString().startsWith("keystore-"))) throw new AssertionError("temporary file leaked"); } return "WalletUtils.writeWalletFile removes temporary file after failed replace"; } throw new AssertionError(id);
    } finally { deleteTree(dir); }
  }

  private static String filesystemDecision(String id) {
    if (id.startsWith("LIST.")) return "Java has no secure directory-list API; Rust rejects symlinks/insecure ownership or mode and reports malformed entries explicitly.";
    if (id.startsWith("IMPORT.")) return "Java generateWalletFile replaces name collisions and has no duplicate-address/force policy; Rust separates duplicate force from overwrite.";
    if (id.startsWith("UPDATE.")) return "Java has no atomic password-update API; Rust uniquely resolves address, refuses symlink swaps, and preserves original bytes on failure.";
    if ("NEW.INJECTED".equals(id)) return "Java Wallet randomness is ambient SecureRandom; Rust supports injected randomness for deterministic verification without weakening production randomness.";
    return "Java non-POSIX permissions are best-effort and do not fsync the directory; Rust returns an explicit WindowsPermissionsBestEffort decision.";
  }

  private static Map<String, Object> filesystem(String id, String basis, String decision) { Map<String,Object> row=map(); row.put("id",id); row.put("kind","filesystem"); row.put("oracle_basis",basis); row.put("decision",decision); return row; }
  private static Map<String, Object> decision(String id,String kind,String basis,String text) { Map<String,Object> row=map();row.put("id",id);row.put("kind",kind);row.put("oracle_basis",basis);row.put("decision",text);return row; }

  private static String deterministicWallet(String kdf, boolean ec) {
    String address=ec?EC_ADDRESS:SM2_ADDRESS;
    String cipher=kdf.equals("scrypt")?"1f578b7cbed5ef355df449ab940dc3b6404c2422bb7a6588feeebeae34f21806":"791a2f36aa1ce0c4822e744ac3a1a8a7982dcb938e9ecf10a52dc038766bb5e8";
    String mac=kdf.equals("scrypt")?"270a215c9f282709e377ed53200dc6c9b25188dc3df04d8648f5c1b6e19b5e2f":"b99fa60132b61365e8ad323dbea6fe4f997616189b2047ccd0ccaceff21698e2";
    String params=kdf.equals("scrypt")?"{\"dklen\":32,\"n\":4096,\"p\":6,\"r\":8,\"salt\":\""+SALT+"\"}":"{\"c\":4096,\"dklen\":32,\"prf\":\"hmac-sha256\",\"salt\":\""+SALT+"\"}";
    return "{\"address\":\""+address+"\",\"crypto\":{\"cipher\":\"aes-128-ctr\",\"cipherparams\":{\"iv\":\""+IV+"\"},\"ciphertext\":\""+cipher+"\",\"kdf\":\""+kdf+"\",\"kdfparams\":"+params+",\"mac\":\""+mac+"\"},\"id\":\""+UUID+"\",\"version\":3}";
  }
  private static byte[] hex(String s){byte[] out=new byte[s.length()/2];for(int i=0;i<out.length;i++)out[i]=(byte)Integer.parseInt(s.substring(i*2,i*2+2),16);return out;}
  private static String hx(byte[] b){StringBuilder out=new StringBuilder();for(byte x:b)out.append(String.format("%02x",x&255));return out.toString();}
  private static Map<String,Object> map(){return new LinkedHashMap<>();}
  private static void deleteTree(Path root)throws Exception{if(root==null||!Files.exists(root))return;try(java.util.stream.Stream<Path>s=Files.walk(root)){for(Path p:(Iterable<Path>)s.sorted(java.util.Comparator.reverseOrder())::iterator)Files.deleteIfExists(p);}}
}

File: tools/keystore/c005_oracle.py
#!/usr/bin/env python3
"""Build and execute the C005 oracle against pinned java-tron crypto classes."""
from __future__ import annotations
import argparse, base64, binascii, hashlib, json, os, sys, tempfile
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
SOURCE = ROOT / "tools/keystore/C005Oracle.java"
OUTPUT = ROOT / "docs/oracles/c005-keystore-fixture-manifest.v1.json"
GRADLE_VERSION = "7.6.4"
JAVA_MAJOR = "1.8"

sys.path.insert(0, str(ROOT / "tools/reference-runner"))
from java_reference_guard import install_java_reference_guard

_injected_java = os.environ.get("C029_JAVA_REFERENCE_ROOT")
_guard_root = Path(_injected_java).resolve().parent if _injected_java else ROOT
SESSION = install_java_reference_guard(_guard_root)
JAVA_TRON = SESSION.tree
JAVA_HOME = SESSION.java_home

DISPATCH = {
    **{f"C005.CROSS.{k}.{e}": "java_rust_cross_open_vectors" for k in ("SCRYPT", "PBKDF2") for e in ("EC", "SM2")},
    "C005.CROSS.PBKDF2.MISSING_DKLEN.EC": "java_rust_cross_open_vectors",
    **{f"C005.ERROR.{x}": "schema_and_decrypt_error_vectors" for x in ("WRONG_PASSWORD", "VERSION", "VERSION_INT_OVERFLOW", "MISSING_CRYPTO", "CIPHER", "KDF", "PRF")},
    **{f"C005.PASSWORD.{x}": "password_vectors" for x in ("WHITESPACE", "BOM", "EMPTY", "SHORT", "MULTILINE", "MALFORMED_UTF8")},
    **{f"C005.{x}": "filesystem_and_atomic_vectors" for x in ("NEW.INJECTED", "IMPORT.DUPLICATE", "IMPORT.OVERWRITE", "LIST.CORRUPT", "LIST.BOM", "LIST.MULTILINE", "LIST.SYMLINK", "LIST.PERMISSIONS", "LIST.OWNERSHIP", "DIRECT.SYMLINK", "DIRECT.UNQUOTED", "DIRECT.CORRUPT", "UPDATE.PASSWORD", "UPDATE.WRONG_PASSWORD", "UPDATE.DUPLICATE", "UPDATE.SYMLINK_SWAP", "WRITE.CLEANUP", "WRITE.MODE", "WINDOWS.LIMIT")},
}

def checked(command: list[str], **kwargs) -> str:
    result = SESSION.run(command, **kwargs)
    if result.returncode:
        raise RuntimeError(result.stderr.decode("utf-8", "replace"))
    return result.stdout.decode().strip()

def authenticate_toolchain() -> None:
    version = checked([str(JAVA_TRON / "gradlew"), "--version"], cwd=JAVA_TRON)
    if f"Gradle {GRADLE_VERSION}" not in version or not any(" ".join(line.split()).startswith(f"JVM: {JAVA_MAJOR}") for line in version.splitlines()):
        raise RuntimeError(f"C005 requires Gradle {GRADLE_VERSION} on Java {JAVA_MAJOR}")

def classpath() -> str:
    build_root = SESSION.work / "c005-gradle-build"
    init = """def c005BuildRoot = new File(%s)
allprojects { p ->
  p.buildDir = new File(c005BuildRoot, p.path == ':' ? 'root' : p.path.substring(1).replace(':', '/'))
  if (p.path == ':crypto') {
    p.afterEvaluate {
      p.tasks.register('c005RuntimeClasspath') {
        doLast { println 'C005_CLASSPATH_B64=' + p.sourceSets.main.runtimeClasspath.asPath.bytes.encodeBase64().toString() }
      }
    }
  }
}
""" % json.dumps(str(build_root))
    with tempfile.NamedTemporaryFile("w", suffix=".gradle", dir=SESSION.work, delete=False) as handle:
        handle.write(init); init_path = Path(handle.name)
    try:
        built = SESSION.gradle(["-I", str(init_path), ":protocol:jar", ":platform:jar", ":common:jar", ":crypto:classes", "-q"])
        if built.returncode:
            raise RuntimeError(built.stderr.decode("utf-8", "replace"))
        output = checked([str(JAVA_TRON / "gradlew"), "-I", str(init_path), ":crypto:c005RuntimeClasspath", "--no-daemon", "--console=plain", "-q"], cwd=JAVA_TRON)
        sentinel = "C005_CLASSPATH_B64="
        frames = [line[len(sentinel):] for line in output.splitlines() if line.startswith(sentinel)]
        if len(frames) != 1 or not frames[0]:
            raise RuntimeError("C005 Gradle classpath frame missing, duplicated, or empty")
        try:
            raw_classpath = base64.b64decode(frames[0], validate=True).decode("utf-8")
        except (binascii.Error, UnicodeDecodeError) as error:
            raise RuntimeError("C005 Gradle classpath frame is malformed") from error
        if not raw_classpath:
            raise RuntimeError("C005 Gradle classpath frame decoded to an empty value")
        for raw in raw_classpath.split(os.pathsep):
            path = Path(raw)
            if not path.exists() and path.is_relative_to(build_root) and path.parts[-2:] == ("resources", "main"):
                path.mkdir(parents=True)
            if not path.exists():
                raise RuntimeError(f"required C005 classpath entry is missing: {path}")
        return raw_classpath
    finally:
        init_path.unlink(missing_ok=True)

def compile_adapter(cp: str) -> Path:
    classes = SESSION.work / "c005-oracle-classes"
    classes.mkdir(parents=True, exist_ok=True)
    compiled = SESSION.run([str(JAVA_HOME / "bin/javac"), "-source", "8", "-target", "8", "-cp", cp, "-d", str(classes), str(SOURCE)], cwd=SESSION.work, classpath=cp)
    if compiled.returncode:
        raise RuntimeError(compiled.stderr.decode("utf-8", "replace"))
    return classes

def invoke(mode: str = "manifest", engine: str | None = None, wallet_json: str | None = None) -> dict:
    authenticate_toolchain(); cp = classpath(); classes = compile_adapter(cp)
    command = [str(JAVA_HOME / "bin/java"), "-cp", os.pathsep.join((str(classes), cp)), "C005Oracle", mode]
    if engine is not None: command.append(engine)
    if wallet_json is not None: command.append(base64.b64encode(wallet_json.encode()).decode())
    return json.loads(checked(command, cwd=SESSION.work, classpath=os.pathsep.join((str(classes), cp))).splitlines()[-1])

def generate() -> dict:
    value = invoke()
    ids = [row["id"] for row in value["vectors"]]
    if len(ids) != 37 or len(ids) != len(set(ids)) or set(ids) != set(DISPATCH):
        raise RuntimeError("Java oracle must execute exactly the 37 exhaustive C005 IDs")
    placeholders = ("placeholder", "see_c005", "not_run", "todo")
    if any(any(word in json.dumps(row).lower() for word in placeholders) for row in value["vectors"]):
        raise RuntimeError("C005 oracle emitted a placeholder vector")
    value["rust_dispatch"] = DISPATCH
    value["vectors_sha256"] = hashlib.sha256(json.dumps(value["vectors"], sort_keys=True, separators=(",", ":")).encode()).hexdigest()
    return value

def encoded(value: dict) -> bytes:
    return (json.dumps(value, indent=2, sort_keys=True) + "\n").encode()

def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--write", action="store_true")
    parser.add_argument("--java-create", choices=("ec", "sm2"))
    parser.add_argument("--java-open", choices=("ec", "sm2"))
    parser.add_argument("--wallet-json")
    args = parser.parse_args()
    if args.java_create:
        print(json.dumps(invoke("java-create", args.java_create), sort_keys=True)); return 0
    if args.java_open:
        if args.wallet_json is None: parser.error("--java-open requires --wallet-json")
        print(json.dumps(invoke("java-open", args.java_open, args.wallet_json), sort_keys=True)); return 0
    expected = encoded(generate())
    if args.write:
        OUTPUT.write_bytes(expected); print(f"wrote {OUTPUT.relative_to(ROOT)} ({len(DISPATCH)} executed vectors)"); return 0
    actual = OUTPUT.read_bytes() if OUTPUT.is_file() else b""
    if actual != expected:
        print("C005 oracle drift; run tools/keystore/c005_oracle.py --write", file=sys.stderr); return 1
    print(f"C005 pinned Java oracle verified: {len(DISPATCH)} executed vectors, sha256 {hashlib.sha256(actual).hexdigest()}")
    return 0
if __name__ == "__main__": raise SystemExit(main())
